### PR TITLE
643 - Open Telemetry

### PR DESCRIPTION
**Pull request by Thuen**
Adding Open Telemetry to Access Management

## Description
Initial PR to implement Open Telemetry in Access Management. 

## Related Issue(s)
- #643 

## Developer/Reviewer Checklist
- [ ] **Your** code builds clean without any errors or warnings
- [ ] No changes to config/appsettings or environment variables created in separate linked PR
- [ ] Manual testing done (required)
- [ ] Relevant integration test added (required for functional changes)
- [ ] Relevant automated test added (required for functional changes)
- [ ] All integration tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Altinn.AccessManagement.Integration/Clients/AuthenticationClient.cs [hunk 1]
```diff
@@ -60,12 +54,14 @@ public async Task<string> RefreshToken()
                 }
                 else
                 {
-                    _logger.LogError("Refreshing JwtToken failed with status code");
+                    activity?.StopWithError(TelemetryEvents.UnexpectedHttpStatusCode(response)); 
+                    
+                    // Review: Original: _logger.LogError("Refreshing JwtToken failed with status code"); Fix: New ActivityType for JwtToken failed
```
Comment (Thuen) [inline src/Altinn.AccessManagement.Integration/Clients/AuthenticationClient.cs:59]: ### Review
Original: _logger.LogError("Refreshing JwtToken failed with status code"); 
New: activity?.StopWithError(TelemetryEvents.UnexpectedHttpStatusCode(response));
Alternate: New ActivityType for JwtToken failed

File: src/Altinn.AccessManagement.Integration/Clients/ProfileClient.cs [hunk 2]
```diff
@@ -44,21 +38,31 @@ public class ProfileClient : IProfileClient
         /// <inheritdoc/>
         public async Task<UserProfile> GetUser(UserProfileLookup userProfileLookup, CancellationToken cancellationToken = default)
         {
-            UriBuilder endpoint = new UriBuilder($"{_settings.ApiProfileEndpoint}internal/user/");
+            // Review: Added try/catch
+            using var activity = TelemetryConfig._activitySource.StartActivity();
+            try
```
Comment (Thuen) [inline src/Altinn.AccessManagement.Integration/Clients/ProfileClient.cs:43]: ### Review
Added try/catch to match other methods

File: src/Altinn.AccessManagement.Integration/DelegationRequestProxy.cs [hunk 3]
```diff
@@ -72,7 +71,9 @@ public async Task<DelegationRequests> GetDelegationRequestsAsync(string who, str
             }
             else
             {
-                _logger.LogError("Getting delegationg requsts from bridge failed with {StatusCode}", response.StatusCode);
+                activity?.StopWithError(TelemetryEvents.UnexpectedHttpStatusCode(response));
+                
+                // Review: Original: _logger.LogError("Getting delegationg requsts from bridge failed with {StatusCode}", response.StatusCode); Fix: SBLBride.RequestFailed
```
Comment (Thuen) [inline src/Altinn.AccessManagement.Integration/DelegationRequestProxy.cs:76]: ### Review
Original: _logger.LogError("Getting delegationg requsts from bridge failed with {StatusCode}", response.StatusCode); 
New: activity?.StopWithError(TelemetryEvents.UnexpectedHttpStatusCode(response));
Alternate: SBLBride.RequestFailed

File: src/Altinn.AccessManagement.Core/Telemetry/TelemetryEvents.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    /// Collection of ActivityEvents specific for Sbl Bridge
+    /// </summary>
+    public static class SblBridge
+    {
+        /// <summary>
+        /// ActivityEvent when Sbl Bridge is unreachable
+        /// </summary>
+        /// <param name="description">Description of reason or other relevant information</param>
+        /// <returns></returns>
+        public static ActivityEvent Unreachable(string description)
```
Comment (andreasisnes) [inline src/Altinn.AccessManagement.Core/Telemetry/TelemetryEvents.cs:35]: :+1:

File: src/Altinn.AccessManagement.Core/Telemetry/TelemetryConfig.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+using System.Diagnostics.CodeAnalysis;
+using System.Diagnostics.Metrics;
+
+namespace Altinn.AccessManagement.Core.Telemetry;
+
+/// <summary>
+/// Holds configuration for OpenTelemetry implementation
+/// </summary>
+[ExcludeFromCodeCoverage]
+public static class TelemetryConfig
```
Comment (andreasisnes) [inline src/Altinn.AccessManagement.Core/Telemetry/TelemetryConfig.cs:11]: TelemetrySources?

File: src/Altinn.AccessManagement.Core/Telemetry/TelemetryConfig.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+{
+    /// <summary>
+    /// Used as source for the current project
+    /// </summary>
+    public static readonly ActivitySource _activitySource = new("Altinn.AccessManagement");
```
Comment (andreasisnes) [inline src/Altinn.AccessManagement.Core/Telemetry/TelemetryConfig.cs:16]: _activitySource, rename ActivitySource?

File: src/Altinn.AccessManagement.Core/Telemetry/TelemetryConfig.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+    /// <summary>
+    /// Used as source for the current project
+    /// </summary>
+    public static readonly Meter _meter = new("Altinn.AccessManagement");
```
Comment (andreasisnes) [inline src/Altinn.AccessManagement.Core/Telemetry/TelemetryConfig.cs:21]: _meter, rename Meter?

**Comment by andreasisnes** [comment]
Setting it to closed after consent from Marius

**Pull request closed without merging**